### PR TITLE
json: New streaming `JSONLoader` (supports GeoJSON)

### PR DESCRIPTION
**Pull request by ibgreen**
See RFC in #376 

Rewritten on top of a `clarinet` fork. Now supports streaming GeoJSON.

### CODE REVIEW

File: modules/experimental/src/index.js [hunk 1]
```diff
@@ -4,11 +4,13 @@ export {writeFile} from './core-addons/write-file-browser';
 export {default as AsyncQueue} from './javascript-utils/async-queue';
```
Comment (Pessimistress) [inline modules/experimental/src/index.js]: Can we just drop the utilities from `experimental` and move them into `core` or `loader-utils`?

File: modules/json/src/json-loader.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+function parseTextSync(text, options) {
+  return JSON.parse(text);
+}
+
+export default {
+  name: 'JSON',
+  extensions: ['json'],
+  testText: null,
```
Comment (Pessimistress) [inline modules/json/src/json-loader.js]: Didn't we have a tester?

File: modules/json/src/json-table-loader.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
+/* global Buffer */
+import {AsyncQueue} from '@loaders.gl/experimental';
+import {TableBatchBuilder, RowTableBatch} from '@loaders.gl/experimental/categories/table';
+import JSONParser from './lib/json-parser';
+
+export default {
```
Comment (Pessimistress) [inline modules/json/src/json-table-loader.js]: I'd prefer to merge this with the `JSONLoader`. This is a new module and it doesn't look like they have any conflicts...

File: modules/json/docs/README.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+- if the filename extension is `.jsonl`
+- if the MIMETYPE is `application/json-seq`
+- if the first value in the file is a number, assume the file is length prefixed.
+
+For data in non-streaming JSON format, the presence of a top-level array will start streaming of objects.
+
+For embedded arrays, a path specifier may need to be supplied (or could look for first array).
+
+### Streaming JSON Formats
```
Comment (Pessimistress) [inline modules/json/docs/README.md]: Does it support GeoJSON feature collection yet?

File: modules/json/src/json-table-loader.js (hunk 3 above)
Comment (ibgreen) [reply modules/json/src/json-table-loader.js]: > I'd prefer to merge this with the `JSONLoader`. This is a new module and it doesn't look like they have any conflicts 

I am not sure that makes sense. Two reasons to keep them separate:

- The `JSONTableLoader` is not a loader for generic JSON objects (unstructured "tree" of data). This is a loader for JSON containing a "top-level" array of JSON objects with identical, relatively flat structure. We still need a generic `JSONLoader`.

- The JSONLoader is 5 lines and extremely lightweight. This loader will be at least 1000 lines, and likely also pull in table batch library from the tables category.

File: modules/json/docs/README.md (hunk 4 above)
Comment (ibgreen) [reply modules/json/docs/README.md]: > Does it support GeoJSON feature collection yet?

No that's an extra wrinkle that I'd like to tackle separately.

File: modules/experimental/src/index.js (hunk 1 above)
Comment (ibgreen) [reply modules/experimental/src/index.js]: > Can we just drop the utilities from experimental and move them into core or loader-utils?

Yes the goal is to get all table related stuff out of experimental. I am thinking that we'll need a table category module for all the table batch / arrow compatibility stuff.

The core module is growing really fast right now and I want to be careful with adding things there.

File: modules/json/src/json-loader.js (hunk 2 above)
Comment (ibgreen) [reply modules/json/src/json-loader.js]: > Didn't we have a tester?

I thought so but can't find it.

**Comment by ibgreen** [comment]
Rewritten on top of a `clarinet` fork. Now also supports streaming GeoJSON.

File: modules/json/docs/README.md (hunk 4 above)
Comment (ibgreen) [reply modules/json/docs/README.md]: Now supports GeoJSON (it simply "streams" the first array (i.e. non-object) it encounters in the JSON object, which would typically be the contents of the "features" field in a GeoJSON file).

File: modules/experimental/src/index.js (hunk 1 above)
Comment (ibgreen) [reply modules/experimental/src/index.js]: These were moved into the `tables` category in previous diffs.

File: docs/whats-new.md [hunk 5]
```diff
@@ -1,5 +1,13 @@
 # What's New
 
+## v2.1 (In Development)
+
+Release Date: Target Feb-Mar, 2019.
+
+### @loaders.gl/json
```
Comment (ibgreen) [inline docs/whats-new.md]: @xintongxia Should we just include the experimental JSONLoader in 2.0?

File: modules/json/src/json-loader.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    ndjson: {stream: true}
+  },
+  mimeTypes: {
+    'application/json': null,
+    'application/json-seq': {stream: true},
+    'application/x-ndjson': {stream: true}
+  },
+  */
+  category: 'table',
+  parse: async (arrayBuffer, options) =>
```
Comment (Pessimistress) [inline modules/json/src/json-loader.js]: Why is a function that calls `parseJSONSync` async?

Comment (ibgreen) [reply modules/json/src/json-loader.js]: The convention is that all loaders have at least an async `parse` function. 

But yes, I suppose we should also offer a `parseSync`.

**Pull request merged**